File: cli/src/core/TopicKBTypes.ts
/**
 * Topic KB — shared type declarations for the topic-centric knowledge base
 * (sub-project 1). Pure type module: no runtime behavior.
 */

/** The four source streams folded into the knowledge base. */
export type SourceType = "summary" | "plan" | "note" | "userfile";

/** A single ingestable source, identified stably and timestamped for ordering. */
export interface SourceRef {
	readonly type: SourceType;
	/** Stable identity: commit hash / plan slug / note id / `path@fingerprint`. */
	readonly id: string;
	/** ISO 8601; used for chronological ordering (parsed to epoch, may carry tz offset). */
	readonly timestamp: string;
	/**
	 * Originating branch for branch-scoped sources (summary / plan / note). The
	 * authoritative input to a topic page's `relatedBranches` — the LLM's
	 * `---RELATEDBRANCHES---` echo is advisory only. Absent for userfiles (repo /
	 * global knowledge, not branch-scoped) and for refs deserialized from pages
	 * written before this field existed.
	 */
	readonly branch?: string;
}

/** High-water mark = the set of already-ingested source IDs, grouped by type. */
export interface ProcessedSet {
	readonly schemaVersion: 1;
	readonly processed: Record<SourceType, string[]>;
}

/** One entry in `topics/index.json`. Drives index-driven routing (sub-project 2). */
export interface TopicIndexEntry {
	readonly stableSlug: string;
	readonly title: string;
	readonly summary: string;
	readonly relatedBranches: string[];
	readonly sourceRefs: SourceRef[];
	readonly lastUpdatedAt: string;
}

/** `topics/index.json` shape. */
export interface TopicIndex {
	readonly schemaVersion: 1;
	readonly topics: TopicIndexEntry[];
}

/** Canonical topic page (`topics/<stableSlug>.json`). Content filled by sub-project 2. */
export interface TopicPage {
	readonly schemaVersion: 1;
	readonly stableSlug: string;
	readonly title: string;
	readonly content: string;
	readonly relatedBranches: string[];
	readonly sourceRefs: SourceRef[];
	readonly lastUpdatedAt: string;
}

File: intellij/src/main/kotlin/ai/jolli/jollimemory/core/TopicKBTypes.kt
package ai.jolli.jollimemory.core

/**
 * Topic KB — shared type declarations for the topic-centric knowledge base.
 * Pure type module: no runtime behavior.
 *
 * Kotlin port of `cli/src/core/TopicKBTypes.ts`. Field order mirrors the TS
 * interfaces so Gson serialization stays byte-stable with the CLI/VS Code
 * `JSON.stringify` output (the `topics/` JSON layer is synced cross-device).
 */

/** The four source streams folded into the knowledge base. */
object SourceType {
    const val SUMMARY = "summary"
    const val PLAN = "plan"
    const val NOTE = "note"
    const val USERFILE = "userfile"

    /** All types, in the canonical order used by [ProcessedSet] and tie-break ranking. */
    val ALL = listOf(SUMMARY, PLAN, NOTE, USERFILE)
}

/**
 * A single ingestable source, identified stably and timestamped for ordering.
 *
 * @property type one of [SourceType]
 * @property id stable identity: commit hash / plan slug / note id / `path@fingerprint`
 * @property timestamp ISO 8601; parsed to epoch for chronological ordering (may carry tz offset)
 * @property branch originating branch for branch-scoped sources (summary/plan/note);
 *   authoritative input to a topic page's `relatedBranches`. Absent for userfiles and
 *   for refs deserialized from pages written before this field existed.
 */
data class SourceRef(
    val type: String,
    val id: String,
    val timestamp: String,
    val branch: String? = null,
)

/** High-water mark = the set of already-ingested source IDs, grouped by type. */
data class ProcessedSet(
    val schemaVersion: Int = 1,
    val processed: Map<String, List<String>> = emptyMap(),
)

/** One entry in `topics/index.json`. Drives index-driven routing. */
data class TopicIndexEntry(
    val stableSlug: String,
    val title: String,
    val summary: String,
    val relatedBranches: List<String> = emptyList(),
    val sourceRefs: List<SourceRef> = emptyList(),
    val lastUpdatedAt: String,
)

/** `topics/index.json` shape. */
data class TopicIndex(
    val schemaVersion: Int = 1,
    val topics: List<TopicIndexEntry> = emptyList(),
)

/** Canonical topic page (`topics/<stableSlug>.json`). */
data class TopicPage(
    val schemaVersion: Int = 1,
    val stableSlug: String,
    val title: String,
    val content: String,
    val relatedBranches: List<String> = emptyList(),
    val sourceRefs: List<SourceRef> = emptyList(),
    val lastUpdatedAt: String,
)
